### PR TITLE
[rb] implement navigation commands with BiDi

### PR DESCRIPTION
**Pull request by titusfortner**
### Description
* creates context manager class, constructed with bridge instance
* pulls page load strategy from capabilities and uses in required methods
* we eventually want to figure out how to store current browsing context as state on the bridge (I think)? For now this always gets the current window handle first
* deprecates `BrowsingContext`

### Considerations
* navigate and reload return `NavigateResult` objects; should I return it as a Struct or just ignore it since we aren't doing anything with it and not promising a return value? 
* Should these BiDi classes be marked private api?
* I want to implement this as `BrowsingContext` instead of creating a new class and deprecating that one; do we need to be backwards compatible with this since we didn't mark it private api?

### Motivation and Context
Ruby implementation of #13995

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added `ContextManager` class to handle navigation commands using BiDi.
- Deprecated `BrowsingContext` class in favor of `ContextManager`.
- Enhanced `BiDiBridge` with new navigation methods: `get`, `go_back`, `go_forward`, and `refresh`.
- Added integration tests for navigation methods.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>bidi.rb</strong><dd><code>Add `ContextManager` to autoload list in BiDi module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/bidi.rb

- Added `ContextManager` to autoload list.

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14094/files#diff-b7660c7b53fec20ce95452b2afdbf306e9c36363e5bf8812863375b7bc96e42e">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>browsing_context.rb</strong><dd><code>Deprecate `BrowsingContext` class in favor of `ContextManager`</code></dd></summary>
<hr>
      
rb/lib/selenium/webdriver/bidi/browsing_context.rb

- Added deprecation warning for `BrowsingContext` class.

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14094/files#diff-1faa7da01133f7c8b2bd65184c6eb95593a1823c8a69d51a8e11c29c47cfa8e7">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>context_manager.rb</strong><dd><code>Implement `ContextManager` class with navigation methods</code>&nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/bidi/context_manager.rb

<li>Created <code>ContextManager</code> class.<br> <li> Implemented navigation methods: <code>navigate</code>, <code>traverse_history</code>, and <br><code>reload</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14094/files#diff-d33b3544b8c2f2fd2da4cae84f30b2bf830dbfb91708995b4fab8f8d6a573ea4">+59/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>bidi_bridge.rb</strong><dd><code>Add navigation methods and `context_manager` to `BiDiBridge`</code></dd></summary>
<hr>
      
rb/lib/selenium/webdriver/remote/bidi_bridge.rb

<li>Added navigation methods: <code>get</code>, <code>go_back</code>, <code>go_forward</code>, and <code>refresh</code>.<br> <li> Introduced private <code>context_manager</code> method.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14094/files#diff-812b66f6ce3688d462c12e405e4808ffc075e0b6b2b9804b236ca977539a46e7">+22/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>navigation_spec.rb</strong><dd><code>Add integration tests for navigation methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/spec/integration/selenium/webdriver/navigation_spec.rb

- Added tests for navigation methods: back, forward, and refresh.

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14094/files#diff-254580087740f237f15e45221c3b7adb90c6fafeca31bedc75c0b977722138cc">+33/-30</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: rb/lib/selenium/webdriver/bidi/context_manager.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        def initialize(bridge)
+          @bridge = bridge
+          @bidi = @bridge.bidi
+          page_load_strategy = bridge.capabilities[:page_load_strategy]
+          @readiness = READINESS_STATE[page_load_strategy]
+        end
+
+        def navigate(url, context_id: nil)
+          context_id ||= @bridge.window_handle
+          puts "READINESS - #{@readiness}"
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/context_manager.rb]: Use logger instead.

File: rb/lib/selenium/webdriver/bidi/context_manager.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+require_relative 'navigate_result'
+require_relative 'browsing_context_info'
+
+module Selenium
+  module WebDriver
+    class BiDi
+      class ContextManager
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/context_manager.rb]: Let's mark this private API.

**Review by p0deje — APPROVED** [review]
> navigate and reload return NavigateResult objects; should I return it as a Struct or just ignore it since we aren't doing anything with it and not promising a return value?

I think it's fine not to return anything so far.

> Should these BiDi classes be marked private api?

Yes, please.

> I want to implement this as BrowsingContext instead of creating a new class and deprecating that one; do we need to be backwards compatible with this since we didn't mark it private api?

I think it should be fine to change it since it was part of BiDi which is not widely used.

File: rb/lib/selenium/webdriver/bidi/context_manager.rb (hunk 1 above)
Comment (titusfortner) [reply rb/lib/selenium/webdriver/bidi/context_manager.rb]: Oh that was just to debug a problem, I'll delete. I wasnt parsing the readiness state properly from capabilities, so it was sending the wrong thing.

File: rb/lib/selenium/webdriver/bidi/context_manager.rb (hunk 2 above)
Comment (titusfortner) [reply rb/lib/selenium/webdriver/bidi/context_manager.rb]: I'll do the same with log manager class

**Comment by aguspe** [comment]
Thank you for your contribution @titusfortner ! Have a great day!

**Pull request merged**